HTML entity parsing hits SegmentedString::pushBack() assert through document.write()

obyknovenius opened the pull request:
#### 818118e729fb78d177e97861bd47b876808979c2
<pre>
HTML entity parsing hits SegmentedString::pushBack() assert through document.write()
<a href="https://bugs.webkit.org/show_bug.cgi?id=268217">https://bugs.webkit.org/show_bug.cgi?id=268217</a>

Reviewed by Chris Dumez and Darin Adler.

When advancing past a single-character substring, we should always mark
such a substring as fully consumed (i.e., set its `length` to `0`),
and add its consumed character number to `m_numberOfCharactersConsumedPriorToCurrentSubstring`.

* LayoutTests/TestExpectations:
* Source/WebCore/platform/text/SegmentedString.cpp:
(WebCore::SegmentedString::advancePastSingleCharacterSubstringWithoutUpdatingLineNumber):

Canonical link: <a href="https://commits.webkit.org/283540@main">https://commits.webkit.org/283540@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ab815e196e5acdb2585effc2987fcf239345c2dd

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/66457 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/45832 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/19078 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/70490 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/17590 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/68575 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/53631 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/17350 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/53301 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/11897 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/69524 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/42237 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/57527 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/33955 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/38908 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/14915 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/15943 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/60812 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/15257 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/72193 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/10414 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/14629 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/60626 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/10446 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/57596 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/60940 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14696 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/8597 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/2207 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/41639 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/42716 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/43899 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/42459 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez reviewed (approved):
Seems reasonable but please give other reviewers 24 hours to comment before landing in case they catch something I didn't.

darinadler commented on LayoutTests/TestExpectations:
```diff
@@ -6876,13 +6876,6 @@ imported/w3c/web-platform-tests/webcodecs/audio-encoder.https.any.html [ Failure
 webkit.org/b/258192 imported/w3c/web-platform-tests/webcodecs/full-cycle-test.https.any.worker.html?h264_avc [ Pass Failure ]
 
-# These tests crash in debug since their import (webkit.org/b/268217).
-[ Debug ] imported/w3c/web-platform-tests/html/syntax/parsing/html5lib_entities01.html?run_type=write_single [ Skip ]
-[ Debug ] imported/w3c/web-platform-tests/html/syntax/parsing/html5lib_html5test-com.html?run_type=write_single [ Skip ]
-[ Debug ] imported/w3c/web-platform-tests/html/syntax/parsing/html5lib_plain-text-unsafe.html?run_type=write_single [ Skip ]
-[ Debug ] imported/w3c/web-platform-tests/html/syntax/parsing/html5lib_tests2.html?run_type=write_single [ Skip ]
-[ Debug ] imported/w3c/web-platform-tests/html/syntax/parsing/html5lib_tests24.html?run_type=write_single [ Skip ]
```
Do these tests really cover this enough? Should we add even more tests for better coverage?

obyknovenius replied:
@darinadler I think these are enough. There's one subtest in each which causes the issue. We could extract it in a separate test but it will contain the same logic. I don't think it's necessary.

darinadler commented on Source/WebCore/platform/text/SegmentedString.cpp:
```diff
@@ -182,18 +182,19 @@ void SegmentedString::advanceAndUpdateLineNumber16()
+    m_numberOfCharactersConsumedPriorToCurrentSubstring += m_currentSubstring.numberOfCharactersConsumed();
     if (m_otherSubstrings.isEmpty()) {
-        m_currentSubstring.length = 0;
+        m_currentSubstring = { };
         m_currentCharacter = 0;
```
This is a case that we’d like to optimize. Merging the code with the normal case may be aesthetically appealing and does fix one specific problem here, but we may be forgoing some possibly-valuable efficiency in what should be the common case by using `updateAdvanceFunctionPointers` instead of `updateAdvanceFunctionPointersForEmptyString`, and even by repeatedly zeroing out `m_currentSubstring`.

I’d like to see more rationale for exactly why this is the best change. Despite the bug we are fixing here, I liked the old version with the early return better and this doesn’t seem like a clear improvement.

obyknovenius replied:
@darinadler You're right about `updateAdvanceFunctionPointersForEmptyString()` and early return. I missed that these are two different methods. But we have to reset `m_currentSubstring`. If we don't do that `SegmentedString::numberOfCharactersConsumed()` will give a wrong result later. That was the reason for new test failures.

obyknovenius committed:
HTML entity parsing hits SegmentedString::pushBack() assert through document.write()
https://bugs.webkit.org/show_bug.cgi?id=268217

Reviewed by Chris Dumez and Darin Adler.

When advancing past a single-character substring, we should always mark
such a substring as fully consumed (i.e., set its `length` to `0`),
and add its consumed character number to `m_numberOfCharactersConsumedPriorToCurrentSubstring`.

* LayoutTests/TestExpectations:
* Source/WebCore/platform/text/SegmentedString.cpp:
(WebCore::SegmentedString::advancePastSingleCharacterSubstringWithoutUpdatingLineNumber):

Canonical link: https://commits.webkit.org/283540@main
